api/client/stats_helpers.go

```diff
@@ -58,7 +60,21 @@ func (s *stats) isKnownContainer(cid string) (int, bool) {
 	return -1, false
 }
 
-func (s *containerStats) Collect(cli client.APIClient, streamStats bool) {
+func (s *containerStats) Collect(cli client.APIClient, streamStats bool, waitFirst *sync.WaitGroup) {
+	var (
+		getFirst       bool
+		previousCPU    uint64
+		previousSystem uint64
+		u              = make(chan error, 1)
+	)
+
+	defer func() {
+		// if error happens and we get nothing of stats, release wait group whatever
+		if !getFirst {
+			waitFirst.Done()
```

Comment:
I don't think so. This defer function is to make sure `waitFirst.Done()` will be called when the function returns, and `getFirst` is used to make sure `waitFirst.Done()` will be called only once, so that `WaitGroup.Wait()` won't block or error.

I would add a `getFirst=true` before this line to make it clearer, though it's OK without.